stream.py

```diff
@@ -0,0 +1,91 @@
+import random
+import requests
+import sys
+import time
+from typing import Callable
+from threading import Thread, Event
+from concurrent.futures import ThreadPoolExecutor
+from uuid import uuid4
+import json
+
+
+## User Code
+seconds = 0
+points_per_batch = 10
+
+def read_source() -> dict:
+    global seconds
+    points = [{'timestamp': {'seconds': seconds, 'nanos': i}, 'value': random.random()} for i in range(points_per_batch)]
+    batch = {'channel': 'streaming-test-channel-0', 'tags': {}, 'points': {'type': 'double', 'double': points}}
+    seconds += 1
+    return batch
+## End user code
+
+
+## Nominal client
+class NominalWriteStream:
+    def __init__(self, source: Callable[[], dict], data_source_id: str, push_freq_sec: float | int) -> None:
+        """Create the write stream, save the stop event."""
+        self.data_source_id = data_source_id
+        self.push_freq_sec = push_freq_sec
+        self.source = source
+        self.sink = f"sink_{uuid4()}.jsonl"
+        self.stop_event = Event()
+        self.pool = ThreadPoolExecutor(max_workers=10)
+        self.thread = Thread(target=self.worker, args=(self.stop_event,))
+        self.thread.start()
+
+    def _write_sink(self, write_req_data: dict) -> None:
+        """Threaded entrypoint to write to the sync in the pool."""
+        time.sleep(0.5) # simulate some network request lag
+        with open(self.sink, "a") as sink:
+            json.dump(write_req_data, sink)
+            sink.write('\n')
+
+    def worker(self, stop_event: Event) -> None:
+        """Worker process."""
+        while not stop_event.is_set():
+            source_data = self.source()
+            write_req_data = {"batches": [source_data], "dataSourceRid": self.data_source_id}
+            self.pool.submit(self._write_sink, write_req_data)
+            time.sleep(self.push_freq_sec)
```

Comment:
you can sleep on the stop event so that the sleep is interrupted if the event is set: `self.event.wait(t)`

Reply:
nice thanks